### PR TITLE
Ensure we aren't writing to a closed leveldb database

### PR DESCRIPTION
**Pull request by eric**
I've been seeing more segfaults recently and I wanted to ensure that we aren't trying to use a leveldb handle after it's been closed.

### CODE REVIEW

**Comment by eric** [comment]
This is one segfault I captured: https://gist.github.com/eric/9e7bc3fbeb84192dc2a0

**Comment by ztellman** [comment]
Honestly, this may just be a `leveldb-jni` sucks situation.  We're moving off it elsewhere, in part due to inexplicable segfaults.

**Comment by eric** [comment]
What are you moving to?

**Comment by ztellman** [comment]
Something less general, effectively a write-once key-value store similar to cdb or SSTables.  I'm not sure what the right alternative in this situation will be.

**Comment by eric** [comment]
Do you have any experience with the java port of leveldb?

**Comment by eric** [comment]
I think this is worth merging even if there's something better to move to eventually.

**Comment by ztellman** [comment]
I agree it's worth merging, and I haven't tried the Java port, though presumably clj-leveldb should be able to target that transparently.

**Comment by aphyr** [comment]
lmdb and rocksdb might also be worth investigating; heard excellent performance reports.

File: src/skuld/db/level.clj [hunk 1]
```diff
@@ -47,15 +53,19 @@
         (swap! count-cache inc))))
 
   (close! [db]
-    (.close ^Closeable level))
+    (swap! running (fn [was]
+                     (when was
+                       (.close ^Closeable level)
```
Comment (aphyr) [inline src/skuld/db/level.clj]: Should probably do (locking running) here; multiple threads could .close concurrently.

Comment (aphyr) [inline src/skuld/db/level.clj]: (assuming .close is actually thread-unsafe)

Comment (eric) [inline src/skuld/db/level.clj]: Okay, I'll just lock db in `close!` too.

**Comment by eric** [comment]
Updated.

**Pull request merged**